File: python-server/models/password_reset.py
from extensions import db, Base
from datetime import datetime, timedelta
import uuid

class PasswordReset(Base):
    __tablename__ = 'password_reset'
    
    id = db.Column(db.Integer, primary_key=True)
    user_id = db.Column(db.Integer, db.ForeignKey('user.id'), nullable=False)
    token = db.Column(db.String(255), unique=True, nullable=False)
    created_at = db.Column(db.DateTime, default=datetime.utcnow)
    expires_at = db.Column(db.DateTime, nullable=False)
    used = db.Column(db.Boolean, default=False)
    
    # Relationship with user
    user = db.relationship('User', backref='password_resets', lazy=True)
    
    @staticmethod
    def generate_token():
        """Generate a unique token for password reset"""
        return str(uuid.uuid4())
    
    @staticmethod
    def create_for_user(user, expires_in_hours=24):
        """Create a new password reset token for a user"""
        token = PasswordReset.generate_token()
        reset = PasswordReset(
            user_id=user.id,
            token=token,
            expires_at=datetime.utcnow() + timedelta(hours=expires_in_hours)
        )
        db.session.add(reset)
        db.session.commit()
        return reset
    
    def is_expired(self):
        """Check if the token has expired"""
        return datetime.utcnow() > self.expires_at
    
    def is_valid(self):
        """Check if the token is valid (not expired and not used)"""
        return not self.used and not self.is_expired()
    
    def use_token(self):
        """Mark the token as used"""
        self.used = True
        db.session.commit()
        return True

File: go-server/models/password_reset.go
package models

import (
	"time"

	"github.com/google/uuid"
	"gorm.io/gorm"
)

type PasswordReset struct {
	ID        uint           `json:"id" gorm:"primaryKey;autoIncrement"`
	UserID    uint           `json:"user_id" gorm:"not null;index" validate:"required"`
	Token     string         `json:"token" gorm:"uniqueIndex;not null;size:255"`
	CreatedAt time.Time      `json:"created_at" gorm:"autoCreateTime"`
	ExpiresAt time.Time      `json:"expires_at" gorm:"not null"`
	Used      bool           `json:"used" gorm:"default:false"`
	DeletedAt gorm.DeletedAt `json:"-" gorm:"index"`

	// Relationships
	User User `json:"user,omitempty" gorm:"foreignKey:UserID"`
}

// GenerateToken generates a unique token for password reset
func GenerateToken() string {
	return uuid.New().String()
}

// CreateForUser creates a new password reset token for a user
func CreateForUser(db *gorm.DB, userID uint, expiresInHours int) (*PasswordReset, error) {
	token := GenerateToken()
	reset := &PasswordReset{
		UserID:    userID,
		Token:     token,
		ExpiresAt: time.Now().Add(time.Duration(expiresInHours) * time.Hour),
	}

	if err := db.Create(reset).Error; err != nil {
		return nil, err
	}

	return reset, nil
}

// IsExpired checks if the token has expired
func (pr *PasswordReset) IsExpired() bool {
	return time.Now().After(pr.ExpiresAt)
}

// IsValid checks if the token is valid (not expired and not used)
func (pr *PasswordReset) IsValid() bool {
	return !pr.Used && !pr.IsExpired()
}

// UseToken marks the token as used
func (pr *PasswordReset) UseToken(db *gorm.DB) error {
	pr.Used = true
	return db.Save(pr).Error
}

// TableName specifies the table name for GORM
func (PasswordReset) TableName() string {
	return "password_reset"
}